WebParagrah: ellipsis

Rusino opened the pull request:
Part of https://github.com/flutter/flutter/issues/172561

Rusino committed:
Ellipsis

Rusino committed:
Removing garbage

Rusino committed:
syntax error

Rusino committed:
Merge branch 'master' into ellipsizing

mdebbar committed:
Merge branch 'master' into ellipsizing

Rusino committed:
Merge branch 'master' into ellipsizing

mdebbar commented on engine/src/flutter/lib/web_ui/lib/src/engine/web_paragraph/paragraph.dart:
```diff
@@ -614,9 +614,19 @@ class PlaceholderSpan extends ParagraphSpan {
   final String text;
+  // We use TextSpan to get metrics from Chrome in many places,
+  // including empty spans (for example when measuring strut) and
+  // ellipsis span (which inherits textDirection from the span it attaches to).
```
> and ellipsis span (which inherits textDirection from the span it attaches to)

This doesn't seem to be accurate. The ellipsis span text direction is derived from the bidi regions of the ellipsis text.

Rusino replied:
The default text direction that we use to shape the ellipsis comes from the attaching span. The default text direction matters, too. The ellipsis text direction comes from the ellipsis itself, of course.

Rusino committed:
Fixing the tests

(apparently, they were broken for a while but VSCode showed me the
outdated correct images and I didn't notice)

Rusino committed:
Merge branch 'master' into ellipsizing

Rusino committed:
Merge branch 'master' into ellipsizing

Rusino committed:
Merge branch 'master' into ellipsizing

mdebbar commented on engine/src/flutter/lib/web_ui/lib/src/engine/web_paragraph/layout.dart:
```diff
@@ -1050,10 +1109,15 @@ abstract class LineBlock {
+  String getText(int start, int end) {
+    assert(false, 'Only implemented for the ellipsis block');
+    return '';
+  }
```
I think this can be deleted now.

Rusino committed:
Fix some style problems

mdebbar commented on engine/src/flutter/lib/web_ui/lib/src/engine/web_paragraph/paint.dart:
```diff
@@ -126,12 +125,27 @@ class TextPaint {
+              // and all the other blocks with the default paragraph direction
+              block is EllipsisBlock
+                  ? block.isLtr
+                  : layout.paragraph.paragraphStyle.textDirection == ui.TextDirection.ltr,
```
Why are we using paragraph direction to draw all blocks, but we use block direction for ellipsis block? The comment above says "what" we are doing but it doesn't say "why".

Rusino committed:
Reacting to code review comments

Rusino committed:
Merge branch 'master' into ellipsizing
